Source/tusdotnet/Stores/InternalFileRep.cs

```diff
@@ -52,9 +52,9 @@ public string ReadFirstLine(bool fileIsOptional = false)
             return sr.ReadLine();
         }
 
-        public FileStream GetStream(FileMode mode, FileAccess access, FileShare share)
+        public FileStream GetStream(FileMode mode, FileAccess access, FileShare share, bool buffered = true, int bufferSize = 4096)
         {
-            return new FileStream(Path, mode, access, share, bufferSize: 4096, useAsync: true);
+            return new FileStream(Path, mode, access, share, bufferSize, FileOptions.SequentialScan | FileOptions.Asynchronous | (buffered ? FileOptions.None : FileOptions.WriteThrough));
```

Comment:
Very nice! Did not know of these file options. They will hopefully improve performance a lot. 👍 I also like the idea of having the file stream do all the write buffering. Have you checked the implementation of file stream to see how it does this? I'm a bit worried that it uses a byte array in .NET (basically `var buffer = new byte[size]`) which would end up in LOH and cause memory issues. Hopefully it of loads the buffering to the OS and this won't be an issue.

Reply:
Nope, unfortunately [it uses a byte array](https://github.com/dotnet/runtime/blob/832a65623d180412fcbaefd68aa57065341db931/src/libraries/System.Private.CoreLib/src/System/IO/FileStream.cs#L19) in .NET 5. (There are [some ongoing major refractorings](https://github.com/dotnet/runtime/pull/47128) that started just 9 days ago in .NET 6 tho that may change things.)
I basically introduced this change to *skip* the buffering done by FileStream using `FileOptions.WriteThrough` and i thought that this would improve performance since we are already buffering using ArrayPool and dont need FileStream's buffer. Unfortunately, `FileOptions.WriteThrough` didnt really improve performance.